getindex returns arrays with correct dims, or scalars

rafaqz opened the pull request:
This standardises the return value of `getindex` to match base julia `Array` behaviour. 

Dimensions indexed with Int are dropped using `reshape`. If all are dropped, a scalar is returned.

rafaqz commented:
The coverage decrease is just inlining error

rafaqz commented:
I've updated this to also use `Base.to_indices` to align indexing with `AbstractArray` behaviour. `eachindex` is required for this to work on vectors.

musm commented:
I guess we need a breaking change with this PR and several others

musm commented:
Should we try to also fix missing trailing  indices?

musm commented:
I guess that is a bit orthogonal however

I think this is good to go

musm commented on src/HDF5.jl:
```diff
@@ -1830,8 +1835,12 @@ function setindex!(dset::HDF5Dataset, x::Number, indices::Union{AbstractRange{In
 
-getindex(dset::HDF5Dataset, I::Union{AbstractRange{Int},Int,Colon}...) = getindex(dset, ntuple(i-> isa(I[i], Colon) ? (1:size(dset,i)) : I[i], length(I))...)
-setindex!(dset::HDF5Dataset, x, I::Union{AbstractRange{Int},Int,Colon}...) = setindex!(dset, x, ntuple(i-> isa(I[i], Colon) ? (1:size(dset,i)) : I[i], length(I))...)
+getindex(dset::HDF5Dataset, I::Union{AbstractRange{Int},Int,Colon}...) = 
```
```suggestion
function getindex(dset::HDF5Dataset, inds...)
   _getindex(dset, Base.to_indices(dset, inds)...)
end
```

musm replied:
And rename `getindex` to `_getindex` on L1751

musm commented on src/HDF5.jl:
```diff
@@ -1830,8 +1835,12 @@ function setindex!(dset::HDF5Dataset, x::Number, indices::Union{AbstractRange{In
-setindex!(dset::HDF5Dataset, x, I::Union{AbstractRange{Int},Int,Colon}...) = setindex!(dset, x, ntuple(i-> isa(I[i], Colon) ? (1:size(dset,i)) : I[i], length(I))...)
+getindex(dset::HDF5Dataset, I::Union{AbstractRange{Int},Int,Colon}...) = 
+    getindex(dset, Base.to_indices(dset, I)...)
+setindex!(dset::HDF5Dataset, x, I::Union{AbstractRange{Int},Int,Colon}...) = 
```
```suggestion
function setindex!(dset::HDF5Dataset, x, inds...)
    setindex!(dset, x, Base.to_indices(dset, inds)...)
end```

musm replied:
This way indexing should work with, e.g. `Int32` 

`d[Int32(1),Int32(1)]`

musm commented:
This is a net improvement, made a couple of comments. 

We should also try to fix the trailing indices. 

I think after this PR and tackling the dropping trailining indices. The array interface should match Base correct?

rafaqz commented:
Trailing indices should already be omitted with `Base.to_indices`? We should probably test it tho. I'll add a few more tests.

rafaqz commented:
So you are right trailing indices are still broken. Unfortunately the base methods that deal with it are very `AbstractArray` specific. So it would involve writing a custom method to drop trailing 1s. 

Some changes made are I cleaned up the indexing code and added `Array()` constructor so you don't have to use `HDF5.read` to get the whole array.

musm commented on src/HDF5.jl:
```diff
@@ -1752,23 +1752,35 @@ write(parent::Union{HDF5File, HDF5Group}, name::String, data::Union{T, AbstractA
+Base.eachindex(::IndexLinear, A::HDF5Dataset) = Base.OneTo(length(A))
+Base.axes(dset::HDF5Dataset) = map(Base.OneTo, size(dset))
+
+getindex(dset::HDF5Dataset, I::Union{AbstractRange{Int},Int,Colon}...) = begin
```
```
function getindex(dset::HDF5Dataset, I...)
    _getindex(dset, Base.to_indices(dset, I)...)
end
```

we want to be able to index with UInt or Int32 , the method restriction is necessary.

musm replied:
I'm also not confident about the bounds checking approach you took, in fact it doesn't actually do any bounds checking. 
We'd need something like
```
@boundscheck checkbounds(A, I...)
```
if you look `hyperslab` is doing to bounds checking right now.  Maybe just leave this to another PR?

musm commented on src/HDF5.jl:
```diff
@@ -1752,23 +1752,35 @@ write(parent::Union{HDF5File, HDF5Group}, name::String, data::Union{T, AbstractA
+
+# Indexing
+
+Array(dset::HDF5Dataset) = read(dset)
```
Unrelated? Why add this?

rafaqz replied:
I'll put it it another PR for clarity.

musm replied:
did you forget to remove this for this PR?

rafaqz replied:
must have undone it and rebased it back in.

musm commented on src/HDF5.jl:
```diff
@@ -1778,29 +1790,38 @@ function _getindex(dset::HDF5Dataset, T::Type, indices::Union{AbstractRange{Int}
 # Write to a subset of a dataset using array slices: dataset[:,:,10] = array
-function setindex!(dset::HDF5Dataset, X::Array, indices::Union{AbstractRange{Int},Int}...)
+
+setindex!(dset::HDF5Dataset, x, I::Union{AbstractRange{Int},Int,Colon}...) = 
```
Same thing here, no need to restrict the `I` type
```
function setindex!(dset::HDF5Dataset, x, I...)
    _setindex!(dset, x, Base.to_indices(dset, I)...)
end
```

rafaqz commented:
I simplified the PR removing the bounds check and `Array()`. 

Dispatch is on `{AbstractRange,Integer,Colon}` so the correct `MethodError` is given on `getindex` instead of showing all the `_getindex` options. 

Also added a few more tests for completeness.

musm commented:
great thanks

musm commented on src/HDF5.jl:
```diff
@@ -1752,23 +1752,32 @@ write(parent::Union{HDF5File, HDF5Group}, name::String, data::Union{T, AbstractA
+Array(dset::HDF5Dataset) = read(dset)
+
+Base.eachindex(::IndexLinear, A::HDF5Dataset) = Base.OneTo(length(A))
+Base.axes(dset::HDF5Dataset) = map(Base.OneTo, size(dset))
```
I don't this this method is needed for this PR

rafaqz replied:
Its used in `to_indices` to convert `Colon` to range:

https://github.com/JuliaLang/julia/blob/master/base/indices.jl#L321

musm replied:
strange mine was working without this definition

musm commented:
a few minor comments

musm commented:
great thanks for this! Major improvement
